### ISSUE TITLE
Goroutine leak on demo server

### ISSUE THREAD

**Issue by alytvynov** [bradfitz/http2#33]
See https://http2.golang.org/goroutines

No issue about this is open and i couldn't find a TODO in the code about this exactly.
Just want to make sure this is known.

**Comment by shurcooL** [comment]
Relevant: https://twitter.com/bradfitz/status/571804060090679296.

**Comment by alytvynov** [comment]
@shurcooL  Ahh, I knew that i've seen him mention this somewhere that stopped me from opening this issue before.

**Comment by bradfitz** [comment]
I will be offline for two weeks starting tomorrow so I won't have time to investigate this soon. Could somebody else?

**Comment by jaredfolkins** [comment]
Hi Everyone,

I've been hacking on this leak for a while (more time than I'd care to admit) and I believe I can give some insight into what is going on.

# Reproduce

The easiest way to reproduce this issue is to head over to demo server [here](https://http2.golang.org/gophertiles?latency=1000)
 and flood the connection with requests. I do this by holding CMD + R in my browser (Chrome or FF). Please note that it could take a full 10s of flooding the connection for it to hang. I tried to write a test to emulate this request but have been unsuccessful thus far (I need to log the actual request from the phyiscal browser and step through it).

You will see that the connection will hang after a couple of seconds. If you now go to the root of the demo site [here](https://http2.golang.org) your connection is now hung. You will have to close/open your browser or wait for your connection to timeout.

If you dig into the codebase you'll find that the buffered `sc.wroteFrameCh` channel eventually blocks. This leads me to believe that syncronization is off between the handler go routines and the serve go routine.

# Digging deeper

Discovering that the simulated latency is causing a sync issue I wanted to check for race conditons.

On a local machine I made sure to set `GORACE=halt_on_error=1` and then ran the demo server with `go run -race h2demo.go`.

If you flood the server at the same location, [here](https://localhost:4430/gophertiles?latency=1000) you actually get the same results. It just hangs. However if you utilize different simulated latency, you can get the server to find a race condition. Flooding the  connection [here](https://localhost:4430/gophertiles?latency=30) utilizting 30ms of latency should get the following output. 

```bash
➜  h2demo git:(master) ✗ go run -race h2demo.go
2015/04/17 10:42:07 Listening on https://localhost:4430/
==================
WARNING: DATA RACE
Read by goroutine 168:
  github.com/bradfitz/http2.(*Framer).WriteData()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/frame.go:463 +0x116
  github.com/bradfitz/http2.(*writeData).writeFrame()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/write.go:89 +0xbc
  github.com/bradfitz/http2.(*serverConn).writeFrameAsync()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:559 +0x8e

Previous write by goroutine 132:
  github.com/bradfitz/http2.(*Framer).WriteData()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/frame.go:463 +0x3c0
  github.com/bradfitz/http2.(*writeData).writeFrame()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/write.go:89 +0xbc
  github.com/bradfitz/http2.(*serverConn).writeFrameAsync()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:559 +0x8e

Goroutine 168 (running) created at:
  github.com/bradfitz/http2.(*serverConn).startFrameWrite()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:794 +0x4c6
  github.com/bradfitz/http2.(*serverConn).scheduleFrameWrite()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:831 +0x389
  github.com/bradfitz/http2.(*serverConn).serve()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:635 +0xa3c
  github.com/bradfitz/http2.(*Server).handleConn()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:277 +0xfb2
  github.com/bradfitz/http2.func·006()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:189 +0xe2
  net/http.(*conn).serve()
      /usr/local/Cellar/go/1.4.2/libexec/src/net/http/server.go:1155 +0x94a

Goroutine 132 (finished) created at:
  github.com/bradfitz/http2.(*serverConn).startFrameWrite()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:794 +0x4c6
  github.com/bradfitz/http2.(*serverConn).scheduleFrameWrite()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:831 +0x389
  github.com/bradfitz/http2.(*serverConn).writeFrame()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:744 +0x8f
  github.com/bradfitz/http2.(*serverConn).serve()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:632 +0x9d3
  github.com/bradfitz/http2.(*Server).handleConn()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:277 +0xfb2
  github.com/bradfitz/http2.func·006()
      /Users/jared.folkins/go/src/github.com/bradfitz/http2/server.go:189 +0xe2
  net/http.(*conn).serve()
      /usr/local/Cellar/go/1.4.2/libexec/src/net/http/server.go:1155 +0x94a
==================
exit status 66
```

We can now see that `startFrameWrite()` Is called twice. I've had the demo server panic on different resources so your results could vary a bit, but the comminality is that `startFrameWrite()` is called twice when it shouldn't be.

# Hypothesis

What I believe is happening is because of the latency there is a rare occurence were the timing can be such that the Go Scheduler allows the serve go routine to set the `sc.writingFrame` boolean to false. Then two other handler go routines call `scheduleFrameWrite()` and bum rush there way past the writingFrame logic block found [here](https://github.com/bradfitz/http2/blob/a962c3ea8967fbf4648846987db8d7970bc9db36/server.go#L811-L813). Ultimately leading to a race condition or a blocked buffered channel.

Thanks,
Jared

**Comment by pabbott0** [comment]
Fix pushed for worst of it. 1 line! There's still another issue with writeHeaders. It's documented in #44.

@jaredfolkins, thanks for all the info. I, too, would not like to admit how many hours I looked at this.

**Comment by jaredfolkins** [comment]
@pabbott0 good effort!

I had actually implemented this myself but I have some problems with it.

From what I can tell `writingFrame` is still able to be accessed by more than one go routine at a time. I was unwilling to submit something that still allowed for a race and could also starve the server with a leak. Eventually leading to a DOS. If we introduce another leaky abstraction and can't understand why, then how can we assume this is a partial fix? We could be be heading in the wrong direction. It appears that the handler for some reason gets orphaned? I wanted @bradfitz's eyeballs on this because I think it may require larger architectural changes. 

If Brad does decide this direction is the one we are headed, something like the following needs to happen. 

* Sync `writingFrame` safely
* the `writeHeaders()` issue that is then introduced is because `handlerDone()` is called but in some cases never returns, try and figure out why that is. 
* Implement `MaxHandlers` (the TODO is at the top of the page)

Thanks,
Jared

**Comment by bradfitz** [comment]
@jaredfolkins, thanks for the repro details. I arrived the same fix as @jaredfolkins this morning while on a flight. I was happy to see Jared's fix was the same as mine upon landing. I want to write some new tests before accepting any PRs, though.

Fixing the goroutines blocked (after RST_STREAM) until the connection is closed is next, once I have tests for that too.

**Comment by pabbott0** [comment]
So the problem with writeHeaders() blocking is actually almost the same as with the writingFrame race. The stateClosed faked frame dispatch in startFrameWrite was breaking assumptions a second time by not sending an ack on wm.done (as is done in writeAsync). This blocks the writeHeaders() select until the connection went away.

**Comment by jaredfolkins** [comment]
@pabbott0 I was wrong. It does look like only the serve() goroutine is accessing the writingFrame value. I must have misread which writeFrame() method was being called and made an inaccurate deduction.  I apologize about that and am sorry to have wasted your time. Nice work.

**Comment by pabbott0** [comment]
Thanks @jaredfolkins. Don't worry, I didn't spend any additional time on it - I was just commenting based on what I did during previous investigation.

In other news, I think I might be wrong about the the writeHeaders(), or that I only fixed one of the problems. Further testing revealed that writeHeaders calls were still getting blocked forever even after I started sending on wm.done in the stateClosed handling code. Looks like *less* of them get leaked, though, so progress was made? Maybe? Still poking at the code...

**Comment by pabbott0** [comment]
Found the last leak (I hope). sc.closeStream calls sc.writeSched.forgetStream which blows away queued frameWriteMsgs. If that deleted frameWriteMsg contains a writeResHeaders and a non-nil chan in done, the writeHeaders call blocks forever.

I'm not sure why writeHeaders is getting passed rws.frameWriteCh when writeChunk calls it. It seems like writeSchedule.forgetStream() should be close()ing any non-nil done channel before clearing the writeQueue's contents, but I don't know how that would interact with a channel that gets reused.

**Comment by bradfitz** [comment]
Fixed by #44
